open_instruct/rejection_sampling/generation.py

```diff
@@ -97,6 +89,56 @@ def main(args: Args, dataset_args: DatasetArgs, gen_args: GenerationArgs):
             include_stop_str_in_output=True,
         ),
     )
+
+    return [
+        {
+            "outputs": [asdict(out) for out in output.outputs],
+            "prompt": output.prompt,
+            "prompt_logprobs": output.prompt_logprobs,
+            "metrics": output.metrics,
+        }
+        for output in outputs
+    ]
+
+
+def main(args: Args, dataset_args: DatasetArgs, gen_args: GenerationArgs):
+
+    ds = load_dataset(dataset_args.dataset_name)
+    if dataset_args.sanity_check:
+        for key in ds:
+            ds[key] = ds[key].select(range(min(dataset_args.sanity_check_size, len(ds[key]))))
+    if dataset_args.dataset_end_idx is None:
+        dataset_args.dataset_end_idx = len(ds[dataset_args.dataset_train_split])
+    for key in ds:
+        ds[key] = ds[key].select(range(dataset_args.dataset_start_idx, dataset_args.dataset_end_idx))
+    pprint([dataset_args, args, gen_args])
+
+    if "gpt-3.5" in args.model_name_or_path or "gpt-4" in args.model_name_or_path:
+        use_openai = True
+    else:
+        use_openai = False
+
+    if not use_openai:
+        tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path)
+
+        ds = ds.map(
+            lambda x: {"prompt_token_ids": tokenizer.apply_chat_template(x["messages"][:-1])},
+            num_proc=multiprocessing.cpu_count(),
+        )
+        prompt_token_ids = ds[dataset_args.dataset_train_split]["prompt_token_ids"]
+        # Generate using vLLM.
+        outputs = generate_with_vllm(args.model_name_or_path, prompt_token_ids, gen_args)
+
+    else:
+        tokenizer = AutoTokenizer.from_pretrained("allenai/llama-3-tulu-2-8b")
+        ds = ds.map(
+            lambda x: {"prompt": tokenizer.apply_chat_template(x["messages"][:-1], tokenize=False)},
+            num_proc=multiprocessing.cpu_count(),
+        )
+        messages = ds[dataset_args.dataset_train_split]["prompt"]
```

Comment:
We should perhaps use openai's native conversation format (i.e., you can send `[item[:-1] for item in ds[dataset_args.dataset_train_split]["messages"]]`?

```python
from openai import OpenAI
client = OpenAI(
    # Defaults to os.environ.get("OPENAI_API_KEY")
)

chat_completion = client.chat.completions.create(
    model="gpt-4o-mini",
    messages=[
        {"role": "user", "content": "Hi! Do you like reading? I really like reading Game of Thrones."},
        {"role": "assistant", "content": "Hi, I do like reading."},
        {"role": "user", "content": "What did I say I like reading?"},
    ]
)
>>> chat_completion
ChatCompletion(id='chatcmpl-9vsvucRqJKLPaH1UpiqAXIkifroVs', choices=[Choice(finish_reason='stop', index=0, logprobs=None, message=ChatCompletionMessage(content='You mentioned that you really like reading "Game of Thrones." It\'s a fascinating series with rich characters and intricate plots! What do you enjoy most about it?', refusal=None, role='assistant', function_call=None, tool_calls=None))], created=1723582886, model='gpt-4o-mini-2024-07-18', object='chat.completion', service_tier=None, system_fingerprint='fp_48196bc67a', usage=CompletionUsage(completion_tokens=31, prompt_tokens=45, total_tokens=76)
```

Reply:
Since I’m using a template to guide the model on how to respond or provide ratings, I prefer to format it myself rather than relying on the native chat completion. So, the entire conversation will be provided as a single "multi-turn prompt" to the model. Although I could create different templates for "user" and "assistant" I find the outcome to be the same. So, I prefer using a single template to save time and maintain conciseness. But, we can discuss this more and happy to change it if you feel we should stick with native :) 

In `api_generate.py`, text will be my template with the chat formatted between user A and user B. 
``` 
response = await self.async_client.chat.completions.create(
                        model=self.config.model,
                        messages=[
                            {"role": "system", "content": "You are a helpful assistant."},
                            {"role": "user", "content": text},
                        ],
                    )

Reply:
Makes sense. Let's stick with the current design for now and we can change it back if we found the native to work notably better.